docs/unified_api.md

```diff
@@ -0,0 +1,105 @@
+   - **Cannot** have 'Content-Type': 'application/json' header
+   - On error, response contains response.error
+ - Watcher API:
+   - **Must** have 'Content-Type': 'application/json' header
+   - On error, response contains response.result === 'error'
+
+Also on the roadmap is the **Informational API Service** that will provide non-critical convenience APIs.
+
+All three APIs should behave consistently.
+  
+## Proposal
+The eWallet already has a well defined API, using HTTP-RPC (rather than REST).
+ - [eWallet Admin API](https://ewallet.staging.omisego.io/api/admin/docs.ui)
+ - [eWallet Client API](https://ewallet.staging.omisego.io/api/client/docs.ui)
+
+We can follow the same model and ensure consistency across all OmiseGo services.
+
+#### eWallet API characteristics
+The API is a collection of HTTP-RPC style method calls in the format
+```
+EWALLET_URL/api/METHOD_FAMILY.method
+```
+where `METHOD_FAMILY` is one of the functional parts of the API e.g. `account`, `transaction`, etc.
+
+Responses contain all data, metadata or errors in the body of the response. This means that HTTP calls always return `200`, even if the result is an error. One exception to this is if an internal server error occurs - in this case it will return `500`
+
+All HTTP calls are `POST` for consistency.
+
+Following this HTTP-RPC style means that the service can be used via websockets as well as HTTP.
+
+Example:
+```
+POST http://plasma-chain.network/api/account.get_balance
+BODY
+{
+    "address": "0x40d6a26bd478e60f97755d62196f0d0f85c1be0d"
+}
+
+RESPONSE 200
+{
+  "version": "1",
+  "success": true,
+  "data": [
+      {
+          "currency": "0000000000000000000000000000000000000000",
+          "amount": 100
+      },
+      {
+          "currency": "1234560000000000000000000000000000000000",
+          "amount": 300000
+      }
+  ]
+}
+
+RESPONSE 200
+{
+  "version": "1",
+  "success": false,
+  "data": {
+    "object": "error",
+    "code": "account:not_found",
+    "description": "Account not found"
+  }
+}
+
+
+RESPONSE 500
+{
+  "version": "1",
+  "success": false,
+  "data": {
+    "object": "error",
+    "code": "server:internal_server_error",
+    "description": "Something went wrong on the server",
+    "messages": {
+      "error_key": "error_reason"
+    }
+  }
+}
+```
+## OmiseGo Plasma API
+An incomplete list of endpoints to demonstrate the format:
+```
+/account.get_balance
+/account.get_utxos
+/account.get_transactions
+/utxo.get_exit_data
+/utxo.get_challenge_data
+/transaction.get
+/transaction.create
+/transaction.submit
+/block.get
+/status
+```
+
+## Architecture
+Suggestion from Robin is to have a sort of ReverseProxy app that sits in front of the other services and routes the calls to the appropriate service e.g. `/api/transaction.submit` goes to the ChildChain and `/api/account.get_utxos` goes to the Watcher.
+
+I think this can work with HTTP calls, although I'm not sure about WebSockets...
```

Comment:
@kevsul this is the websocket implementation guideline https://github.com/omisego/ewallet/blob/master/docs/guides/ewallet_api_websockets.md

ewallet is using a customized Elixir Phoenix websocket socket base for authentication btw.

Reply:
Thanks @jarindr 
The issue is about the ReverseProxy app though. 
With HTTP the proxy can easily route on the endpoint and redirect to the appropriate app. 
But with WebSockets I imagine that the proxy would have to parse each message to get the action, then route on that. It's probably not all that different to HTTP, but just seems more cumbersome.
I also don't know how WebSocket events would work if there's a proxy server in the middle.

Tagging @robinclart here, because I'm not sure exactly what he had in mind with the proxy server - I may be misunderstanding the intention.

Reply:
oh, it is under architecture 🤣 my bad. Anyway about reverse proxy websocket is possible and not that different from normal http calls since the handshake happen with http upgrade header anyway.
for ref we can use: https://www.nginx.com/blog/websocket-nginx/ as guideline for websocket proxy.
- About reverse proxy, are we talking about devop level or app level (nginx or phoenix)

Reply:
Thank you @kevsul, I like proposition very much, especially simplicity and not fixing to the HTTP.
Great job :exclamation:

Reply:
more note: if the reverse proxy is on the app level, on @omisego/wallet we already have an elixir app doing that https://github.com/omisego/ewallet/tree/master/apps/url_dispatcher which can be reuse easily in this case. cc. @T-Dnzt